lightningd/peer_htlcs.c

```diff
@@ -940,6 +940,16 @@ static bool htlc_accepted_hook_deserialize(struct htlc_accepted_hook_payload *re
 				      " hook: %.*s",
 				      failmsgtok->end - failmsgtok->start,
 				      buffer + failmsgtok->start);
+		} else if ((failoniontok = json_get_member(buffer, toks,
+							       "failure_onion"))) {
+			failonion = json_tok_bin_from_hex(NULL, buffer, failoniontok);
+			if (!failonion)
+				fatal("Bad failure_onion for htlc_accepted"
+				      " hook: %.*s",
+				      failoniontok->end -  failoniontok->start,
+				      buffer + failoniontok->start);
+			fail_in_htlc(hin, take(new_onionreply(tmpctx, failonion)));
+			return false;
```

Comment:
The only difference between `failure_onion` and `failure_msg` is that the latter gets serialized with a prefix and HMAC'd. Would it make sense to pull `local_fail_in_htlc` into this function (basically just the `create_onionreply` call) and handle both cases the same way? That'd mean that we take an eventual `failure_msg`, serialize it using the `create_onionreply` and then store it in `failure_onion`. 

Also I think that `failure_onion` needs to take precedence over `failure_msg`, since more work went into building that one. Ideally we'd panic if both are set, but since this is a hook call, our options are limited, so logging an error is likely the best we can do here.

Reply:
What about this? https://github.com/fiatjaf/lightning/blob/9af91a8493be3eb3b41d422973f2faf6209a02d2/lightningd/peer_htlcs.c#L930-L988

It's not cool, but I think it is clear. And once the deprecated branch is removed it will be much better.